File: src/main/java/com/mudhut/nudge/users/controllers/UserController.java
package com.mudhut.nudge.users.controllers;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import com.mudhut.nudge.users.entities.User;
import com.mudhut.nudge.users.models.AuthResponse;
import com.mudhut.nudge.users.models.ForgotPasswordRequest;
import com.mudhut.nudge.users.models.LoginRequest;
import com.mudhut.nudge.users.models.RegisterRequest;
import com.mudhut.nudge.users.models.ResetPasswordRequest;
import com.mudhut.nudge.users.services.ForgotPasswordService;
import com.mudhut.nudge.users.services.LoginService;
import com.mudhut.nudge.users.services.RegistrationService;
import com.mudhut.nudge.users.services.UserService;
import com.mudhut.nudge.users.services.VerificationService;
import com.mudhut.nudge.utils.models.GeneralRequestResponse;

import jakarta.validation.Valid;

@RestController
@RequestMapping("/api/v1/auth")
public class UserController {

    @Autowired
    private UserService userService;

    @Autowired
    private LoginService loginService;

    @Autowired
    private RegistrationService registrationService;

    @Autowired
    private ForgotPasswordService forgotPasswordService;

    @Autowired
    private VerificationService verificationService;

    @PostMapping("/register")
    public ResponseEntity<User> registerUser(@Valid @RequestBody RegisterRequest request) {
        return ResponseEntity.ok(registrationService.createUser(request));
    }

    @PostMapping("/login")
    public ResponseEntity<AuthResponse> authenticateUser(@Valid @RequestBody LoginRequest request) {
        return ResponseEntity.ok(loginService.authenticateUser(request));
    }

    @PostMapping("/verify-email")
    public ResponseEntity<?> verifyEmail(@RequestParam String token) {
        verificationService.verifyEmail(token);
        return ResponseEntity.ok().build();
    }

    @PostMapping("/verify-phone")
    public ResponseEntity<?> verifyPhone(@RequestParam String code) {
        verificationService.verifyPhone(code);
        return ResponseEntity.ok().build();
    }

    @PostMapping("/forgot-password")
    public ResponseEntity<GeneralRequestResponse> forgotPassword(@Valid @RequestBody ForgotPasswordRequest request) {
        forgotPasswordService.initiateForgotPassword(request.getEmail());
        GeneralRequestResponse response = new GeneralRequestResponse(
                "An email has been sent to you with password reset instructions");
        return ResponseEntity.ok(response);
    }

    @PostMapping("/reset-password")
    public ResponseEntity<GeneralRequestResponse> resetPassword(@Valid @RequestBody ResetPasswordRequest request) {
        forgotPasswordService.resetPassword(request);
        GeneralRequestResponse response = new GeneralRequestResponse(
                "Your password has been reset successfully");
        return ResponseEntity.ok(response);
    }
}

File: src/main/kotlin/com/mudhut/nudge/users/controllers/UserController.kt
package com.mudhut.nudge.users.controllers

import com.mudhut.nudge.users.models.*
import com.mudhut.nudge.users.services.ForgotPasswordService
import com.mudhut.nudge.users.services.GoogleAuthService
import com.mudhut.nudge.users.services.LoginService
import com.mudhut.nudge.users.services.LogoutService
import com.mudhut.nudge.users.services.RegistrationService
import com.mudhut.nudge.users.services.UserService
import com.mudhut.nudge.users.services.VerificationService
import com.mudhut.nudge.utils.models.GeneralRequestResponse
import jakarta.servlet.http.HttpServletRequest
import jakarta.validation.Valid
import org.springframework.http.ResponseEntity
import org.springframework.security.core.Authentication
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/api/v1/auth")
class UserController(
    private val userService: UserService,
    private val loginService: LoginService,
    private val registrationService: RegistrationService,
    private val forgotPasswordService: ForgotPasswordService,
    private val verificationService: VerificationService,
    private val googleAuthService: GoogleAuthService,
    private val logoutService: LogoutService,
) {

    @PostMapping("/register")
    fun registerUser(@Valid @RequestBody request: RegisterRequest): ResponseEntity<UserResponse> =
        ResponseEntity.ok(UserResponse.from(registrationService.createUser(request)))

    @PostMapping("/login")
    fun authenticateUser(@Valid @RequestBody request: LoginRequest): ResponseEntity<AuthResponse> =
        ResponseEntity.ok(loginService.authenticateUser(request))

    @PostMapping("/google")
    fun googleAuth(@Valid @RequestBody request: GoogleAuthRequest): ResponseEntity<AuthResponse> =
        ResponseEntity.ok(googleAuthService.authenticate(request.idToken!!))

    @PostMapping("/verify-email")
    fun verifyEmail(@RequestParam token: String): ResponseEntity<Any> {
        verificationService.verifyEmail(token)
        return ResponseEntity.ok().build()
    }

    @PostMapping("/verify-phone")
    fun verifyPhone(@RequestParam code: String): ResponseEntity<Any> {
        verificationService.verifyPhone(code)
        return ResponseEntity.ok().build()
    }

    @PostMapping("/forgot-password")
    fun forgotPassword(@Valid @RequestBody request: ForgotPasswordRequest): ResponseEntity<GeneralRequestResponse> {
        forgotPasswordService.initiateForgotPassword(request.email!!)
        return ResponseEntity.ok(
            GeneralRequestResponse("An email has been sent to you with password reset instructions")
        )
    }

    @PostMapping("/reset-password")
    fun resetPassword(@Valid @RequestBody request: ResetPasswordRequest): ResponseEntity<GeneralRequestResponse> {
        forgotPasswordService.resetPassword(request)
        return ResponseEntity.ok(
            GeneralRequestResponse("Your password has been reset successfully")
        )
    }

    @PostMapping("/logout")
    fun logout(authentication: Authentication, request: HttpServletRequest): ResponseEntity<Void> {
        logoutService.logout(authentication.name, request.getHeader("Authorization"))
        return ResponseEntity.ok().build()
    }
}
